modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,129 @@
+(ns metabase.driver.presto-jdbc
+    "Presto JDBC driver. See https://prestodb.io/docs/current/ for complete dox."
+    (:require [clojure.java.jdbc :as jdbc]
+              [clojure.set :as set]
+              [clojure.tools.logging :as log]
+              [metabase.db.spec :as db.spec]
+              [metabase.driver :as driver]
+              [metabase.driver.presto-common :as presto-common]
+              [metabase.driver.sql-jdbc.common :as sql-jdbc.common]
+              [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
+              [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
+              [metabase.driver.sql-jdbc.execute.legacy-impl :as legacy]
+              [metabase.driver.sql-jdbc.sync :as sql-jdbc.sync]
+              [metabase.driver.sql-jdbc.sync.describe-database :as sql-jdbc.describe-database]
+              [metabase.util.i18n :refer [trs]])
+    (:import [java.sql Connection ResultSet]))
+
+(driver/register! :presto-jdbc, :parent #{:presto-common :sql-jdbc ::legacy/use-legacy-classes-for-read-and-set})
+
+;;; Presto API helpers
+
+(defmethod sql-jdbc.sync/database-type->base-type :presto-jdbc [_ field-type]
+  ;; seems like we can just delegate to the Presto implementation for this?
+  (presto-common/presto-type->base-type (if (keyword? field-type) (name field-type) field-type)))
+
+(defn- jdbc-spec
+  [{:keys [host port catalog]
+    :or   {host "localhost", port 5432, catalog ""}
+    :as   opts}]
+  (-> (merge
+       {:classname                     "io.prestosql.jdbc.PrestoDriver"
+        :subprotocol                   "presto"
+        :subname                       (db.spec/make-subname host port catalog)}
+       (dissoc opts :host :port :db :catalog))
+      sql-jdbc.common/handle-additional-options))
+
+(defmethod sql-jdbc.conn/connection-details->spec :presto-jdbc
+  [_ {ssl? :ssl, :as details-map}]
+  (let [props (-> details-map
+                  (update :port (fn [port]
+                                    (if (string? port)
+                                      (Integer/parseInt port)
+                                      port)))
+                  (assoc :SSL ssl?)
+                  (dissoc :ssl))]
+       (jdbc-spec props)))
+
+(defn- have-select-privilege?
+  "Checks whether the connected user has permission to select from the given `table-name`, in the given `schema`.
+  Adapted from the legacy Presto driver implementation."
+  [driver conn schema table-name]
+  (try
+   (let [sql (sql-jdbc.describe-database/simple-select-probe-query driver schema table-name)]
+        ;; if the query completes without throwing an Exception, we can SELECT from this table
+        (jdbc/reducible-query {:connection conn} sql)
```

Comment:
do you need to do any ssh-tunnel work at these call sites or when getting the details?

Reply:
The scope of this has sort of blown up so much that I was thinking of leaving tunnel support out of this round.  But yeah, if it gets added then that will have to be accounted for in many places.